feature/experimental/p4rt/ate_tests/google_discovery_protocol_packetout_test/gdp_test.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package google_discovery_protocol_packetout_test
+
+import (
+	"fmt"
+	"net"
+	"testing"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	p4_v1 "github.com/p4lang/p4runtime/go/p4/v1"
+)
+
+type GDPPacketIO struct {
+	PacketIO
+	IngressPort string
+}
+
+// packetGDPRequestGet generates PacketOut payload for GDP packets.
+func packetGDPRequestGet(t *testing.T) []byte {
+	t.Helper()
+	buf := gopacket.NewSerializeBuffer()
+	opts := gopacket.SerializeOptions{
+		FixLengths:       true,
+		ComputeChecksums: true,
+	}
+	pktEth := &layers.Ethernet{
+		SrcMAC: net.HardwareAddr{0x00, 0xAA, 0x00, 0xAA, 0x00, 0xAA},
+		// GDP MAC is 00:0A:DA:F0:F0:F0
+		DstMAC:       net.HardwareAddr{0x00, 0x0A, 0xDA, 0xF0, 0xF0, 0xF0},
+		EthernetType: gdpInLayers,
+	}
+	payload := []byte{}
+	payLoadLen := 64
+	for i := 0; i < payLoadLen; i++ {
+		payload = append(payload, byte(i))
+	}
+	gopacket.SerializeLayers(buf, opts,
+		pktEth, gopacket.Payload(payload),
+	)
+	return buf.Bytes()
+}
+
+// GetPacketOut generates PacketOut message with payload as GDP.
+func (gdp *GDPPacketIO) GetPacketOut(t *testing.T, portID uint32, submitIngress bool) []*p4_v1.PacketOut {
+	packets := []*p4_v1.PacketOut{}
+	packet := &p4_v1.PacketOut{
+		Payload: packetGDPRequestGet(t),
+		Metadata: []*p4_v1.PacketMetadata{
+			{
+				MetadataId: uint32(1), // "egress_port"
+				Value:      []byte(fmt.Sprint(portID)),
+			},
+		},
+	}
+	if submitIngress {
+		packet.Metadata = append(packet.Metadata,
```

Comment:
Do we need to confirm the correctness of the number of bits of the metadata fields after appending them? To check if we would be needing any unused_padding?

Reply:
I think the field is ordered by MetadataId and the value is added based on the MetadataId. Also the value fit the 1-bit which defined in the p4info file. So I think we don't need unused_pad in this case